### PR TITLE
perf(rust): Removed some clones

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR removes unnecessary clones during planning.

The idea is to replace `AsRef<[Expr]>` by `Into<Vec<Expr>>`. The main benefits are:

* API no longer requires passing by ref e.g. (`.select(&[col("a")])) -> .select([col("a")]))`)
* Internals no longer need to clone when not needed

### CODE REVIEW

File: polars/polars-lazy/src/frame/mod.rs [hunk 1]
```diff
@@ -870,20 +854,16 @@ impl LazyFrame {
...
     }
 
     /// Similar to groupby, but order of the DataFrame is maintained.
-    pub fn groupby_stable<E: AsRef<[IE]>, IE: Into<Expr> + Clone>(self, by: E) -> LazyGroupBy {
-        let keys = by
-            .as_ref()
-            .iter()
-            .map(|e| e.clone().into())
-            .collect::<Vec<_>>();
+    pub fn groupby_stable<E: IntoIterator<Item = IE>, IE: Into<Expr>>(self, by: E) -> LazyGroupBy {
```
Comment (jorgecarleitao) [inline polars/polars-lazy/src/frame/mod.rs:865]: this set of APIs is not aligned with the rest of them since they contain an `IE: Into<Expr>`. This change allows users to not have to create a slice to them be iterated by reference (and Expr cloned), instead allowing passing something that iterates by value

File: polars/polars-lazy/src/frame/mod.rs [hunk 2]
```diff
@@ -907,7 +887,12 @@ impl LazyFrame {
     ///         .left_join(other, col("foo"), col("bar"))
     /// }
     /// ```
-    pub fn left_join<E: Into<Expr>>(self, other: LazyFrame, left_on: E, right_on: E) -> LazyFrame {
+    pub fn left_join<L: Into<Expr>, R: Into<Expr>>(
```
Comment (jorgecarleitao) [inline polars/polars-lazy/src/frame/mod.rs:890]: this is a drive-by - before, lhs and rhs had to be of the same type. However, if we want to support `Into<Expr>`, we should allow independent types to be passed to each side.

**Comment by stinodego** [comment]
I am closing this because the last update was almost a year ago.

Feel free to rebase and reopen this if you want to finish the PR!
